apps/prairielearn/src/pages/instructorInstanceAdminPublishing/components/PublishingForm.tsx

```diff
@@ -0,0 +1,615 @@
+        ) {
+          updatedStartDate = oneWeekLater;
+        }
+
+        if (originalEndDate && now <= originalEndDate) {
+          // Try to re-use the original end date if it is in the future.
+          updatedEndDate = DateToPlainDateTime(originalEndDate, courseInstance.display_timezone);
+        } else if (
+          currentEndDate === null ||
+          Temporal.PlainDateTime.compare(currentEndDate, nowTemporal) <= 0 ||
+          Temporal.PlainDateTime.compare(updatedStartDate!, currentEndDate) >= 0
+        ) {
+          updatedEndDate = eighteenWeeksLater;
+        }
+        break;
+      }
+      case 'published': {
+        if (originalStartDate && now >= originalStartDate) {
+          // Try to re-use the original start date if it is in the past.
+          updatedStartDate = DateToPlainDateTime(
+            originalStartDate,
+            courseInstance.display_timezone,
+          );
+        } else if (
+          currentStartDate === null ||
+          // If the current start date is in the future, set it to now.
+          Temporal.PlainDateTime.compare(currentStartDate, nowTemporal) > 0
+        ) {
+          updatedStartDate = nowTemporal;
+        }
+
+        if (originalEndDate && now <= originalEndDate) {
+          // Try to re-use the original end date if it is in the future.
+          updatedEndDate = DateToPlainDateTime(originalEndDate, courseInstance.display_timezone);
+        } else if (
+          currentEndDate === null ||
+          // If the current end date is in the past, set it to 18 weeks from now.
+          Temporal.PlainDateTime.compare(currentEndDate, nowTemporal) < 0
+        ) {
+          updatedEndDate = eighteenWeeksLater;
+        }
+        break;
+      }
+    }
+    setValue('startDate', updatedStartDate === null ? '' : updatedStartDate.toString());
+    setValue('endDate', updatedEndDate === null ? '' : updatedEndDate.toString());
+  };
+
+  const onSubmit = async (data: PublishingFormValues) => {
+    if (!canEdit) return;
+
+    setIsSubmitting(true);
+    try {
+      const requestBody = {
+        __csrf_token: csrfToken,
+        __action: 'update_access_control',
+        accessControl: {
+          startDate: data.startDate === '' ? null : data.startDate,
+          endDate: data.endDate === '' ? null : data.endDate,
+        },
+        orig_hash: origHash,
+      };
+
+      const response = await fetch(window.location.pathname, {
+        method: 'POST',
+        headers: {
+          'Content-Type': 'application/json',
+        },
+        body: JSON.stringify(requestBody),
+      });
+
+      if (response.ok) {
+        window.location.reload();
+        return;
+      }
+
+      const errorData = await response.json();
+      if (errorData.message) {
+        setErrorMessage(errorData.message);
+      }
+    } catch (error) {
+      console.error('Error updating access control:', error);
+      setErrorMessage('Failed to update access control. Please try again.');
```

Comment:
We set error messages here but never clear them if there's a subsequent successful submission. I guess this is OK because we reload the page in that case?

Reply:
That's correct, a successful submission reloads the page.